File: sdks/haima-py/src/haima/x402.py
"""x402 protocol helpers — header parsing, middleware, and auto-pay."""

from __future__ import annotations

import base64
import json
import logging
from typing import Optional

import httpx

from haima.types import PaymentPolicy, PaymentScheme
from haima.wallet import HaimaWallet

logger = logging.getLogger("haima.x402")


def parse_payment_required(header_value: str) -> list[PaymentScheme]:
    """Parse a base64-encoded PAYMENT-REQUIRED header into scheme requirements."""
    decoded = json.loads(base64.b64decode(header_value))
    schemes = decoded.get("schemes", [])
    return [PaymentScheme(**s) for s in schemes]


def parse_payment_response(header_value: str) -> dict:
    """Parse a base64-encoded PAYMENT-RESPONSE header."""
    return json.loads(base64.b64decode(header_value))


class X402Middleware:
    """HTTP transport middleware that auto-handles x402 payment flows.

    Wraps httpx to intercept 402 responses, evaluate against policy,
    sign with wallet, and retry — transparent to the caller.
    """

    def __init__(
        self,
        wallet: HaimaWallet,
        policy: Optional[PaymentPolicy] = None,
        auto_approve: bool = True,
    ):
        self.wallet = wallet
        self.policy = policy or PaymentPolicy()
        self.auto_approve = auto_approve
        self._session_spend = 0
        self._client = httpx.AsyncClient()

    async def request(
        self,
        method: str,
        url: str,
        **kwargs,
    ) -> httpx.Response:
        """Make an HTTP request with automatic x402 handling."""
        response = await self._client.request(method, url, **kwargs)

        if response.status_code != 402:
            return response

        # Parse 402 response
        pr_header = response.headers.get("payment-required")
        if not pr_header:
            return response

        schemes = parse_payment_required(pr_header)
        if not schemes:
            return response

        # Select first compatible scheme (EVM "exact" only)
        scheme = next((s for s in schemes if s.scheme == "exact"), None)
        if not scheme:
            logger.warning("No compatible payment scheme found")
            return response

        # Evaluate against policy
        micro_credits = int(scheme.amount)
        decision = self.policy.evaluate(micro_credits)

        if decision.value == "denied":
            logger.warning("Payment denied by policy: %d μc", micro_credits)
            return response

        if decision.value == "requires_approval" and not self.auto_approve:
            logger.info("Payment requires approval: %d μc", micro_credits)
            return response

        # Check session spend cap
        if self._session_spend + micro_credits > self.policy.session_spend_cap:
            logger.warning("Session spend cap exceeded")
            return response

        # Sign and retry
        signature = self.wallet.sign_payment_header(
            scheme=scheme.scheme,
            network=scheme.network,
            resource_url=url,
            amount=scheme.amount,
            recipient=scheme.recipient,
        )

        headers = dict(kwargs.get("headers", {}))
        headers["payment-signature"] = signature
        kwargs["headers"] = headers

        retry_response = await self._client.request(method, url, **kwargs)
        if retry_response.status_code == 200:
            self._session_spend += micro_credits
            logger.info("Payment settled: %d μc (session total: %d μc)", micro_credits, self._session_spend)

        return retry_response

    async def get(self, url: str, **kwargs) -> httpx.Response:
        return await self.request("GET", url, **kwargs)

    async def post(self, url: str, **kwargs) -> httpx.Response:
        return await self.request("POST", url, **kwargs)

    def reset_session(self) -> None:
        self._session_spend = 0

    async def close(self) -> None:
        await self._client.aclose()

File: sdks/haima-ts/src/x402.ts
/**
 * x402 protocol helpers — header parsing and auto-pay middleware.
 */

import type { PaymentScheme, PaymentPolicy } from "./types.js";
import { DEFAULT_POLICY, evaluatePolicy } from "./types.js";
import { HaimaWallet } from "./wallet.js";

/** Parse a base64-encoded PAYMENT-REQUIRED header. */
export function parsePaymentRequired(headerValue: string): PaymentScheme[] {
  const decoded = JSON.parse(atob(headerValue));
  return decoded.schemes ?? [];
}

/** Parse a base64-encoded PAYMENT-RESPONSE header. */
export function parsePaymentResponse(headerValue: string): {
  tx_hash: string;
  network: string;
  settled: boolean;
} {
  return JSON.parse(atob(headerValue));
}

/**
 * HTTP fetch wrapper that auto-handles x402 payment flows.
 *
 * Intercepts 402 responses, evaluates against policy,
 * signs with wallet, and retries — transparent to the caller.
 */
export class X402Middleware {
  private wallet: HaimaWallet;
  private policy: PaymentPolicy;
  private autoApprove: boolean;
  private sessionSpend = 0;

  constructor(
    wallet: HaimaWallet,
    policy?: Partial<PaymentPolicy>,
    autoApprove = true,
  ) {
    this.wallet = wallet;
    this.policy = { ...DEFAULT_POLICY, ...policy };
    this.autoApprove = autoApprove;
  }

  async fetch(url: string, init?: RequestInit): Promise<Response> {
    const response = await fetch(url, init);

    if (response.status !== 402) {
      return response;
    }

    const prHeader = response.headers.get("payment-required");
    if (!prHeader) return response;

    const schemes = parsePaymentRequired(prHeader);
    if (!schemes.length) return response;

    // Select first compatible scheme (EVM "exact" only)
    const scheme = schemes.find((s) => s.scheme === "exact");
    if (!scheme) return response;

    const microCredits = parseInt(scheme.amount, 10);
    const decision = evaluatePolicy(this.policy, microCredits);

    if (decision === "denied") return response;
    if (decision === "requires_approval" && !this.autoApprove) return response;

    // Check session spend cap
    if (this.sessionSpend + microCredits > this.policy.sessionSpendCap) {
      return response;
    }

    // Sign and retry
    const signature = await this.wallet.signPaymentHeader({
      scheme: scheme.scheme,
      network: scheme.network,
      resourceUrl: url,
      amount: scheme.amount,
      recipient: scheme.recipient,
    });

    const headers = new Headers(init?.headers);
    headers.set("payment-signature", signature);

    const retryResponse = await fetch(url, { ...init, headers });
    if (retryResponse.ok) {
      this.sessionSpend += microCredits;
    }
    return retryResponse;
  }

  resetSession(): void {
    this.sessionSpend = 0;
  }
}
